x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ServiceAccountService.java

```diff
@@ -0,0 +1,155 @@
+     * @return An unvalidated token object.
+     */
+    public static ServiceAccountToken tryParseToken(SecureString token) {
+        try {
+            if (token == null) {
+                return null;
+            }
+            return doParseToken(token);
+        } catch (IOException e) {
+            logger.debug("Cannot parse possible service account token", e);
+            return null;
+        }
+    }
+
+    public void authenticateWithToken(ServiceAccountToken token, ThreadContext threadContext, String nodeName,
+                                      ActionListener<Authentication> listener) {
+
+        if (ElasticServiceAccounts.NAMESPACE.equals(token.getAccountId().namespace()) == false) {
+            logger.debug("only [{}] service accounts are supported, but received [{}]",
+                ElasticServiceAccounts.NAMESPACE, token.getAccountId().asPrincipal());
+            listener.onResponse(null);
+            return;
+        }
+
+        final ServiceAccount account = ACCOUNTS.get(token.getAccountId().serviceName());
+        if (account == null) {
+            logger.debug("the [{}] service account does not exist", token.getAccountId().asPrincipal());
+            listener.onFailure(null);
+            return;
+        }
+
+        if (serviceAccountsCredentialStore.authenticate(token)) {
+            listener.onResponse(success(account, token, nodeName));
+        } else {
+            final ParameterizedMessage message = new ParameterizedMessage(
+                "failed to authenticate service account [{}] with token name []",
+                token.getAccountId().asPrincipal(),
+                token.getTokenName());
+            logger.debug(message);
+            throw new ElasticsearchSecurityException(message.getFormattedMessage());
+        }
+    }
+
+    public void getRoleDescriptor(Authentication authentication, ActionListener<RoleDescriptor> listener) {
+        assert isServiceAccount(authentication) : "authentication is not for service account";
+
+        final ServiceAccountId accountId = ServiceAccountId.fromPrincipal(authentication.getUser().principal());
+        final ServiceAccount account = ACCOUNTS.get(accountId.serviceName());
+        if (account == null) {
+            listener.onFailure(new ElasticsearchSecurityException(
+                "cannot load role for service account [" + accountId.asPrincipal() + "] - no such service account"
+            ));
+            return;
+        }
+        listener.onResponse(account.roleDescriptor());
+    }
+
+    private Authentication success(ServiceAccount account, ServiceAccountToken token, String nodeName) {
+        final User user = account.asUser();
+        final Authentication.RealmRef authenticatedBy = new Authentication.RealmRef(REALM_NAME, REALM_TYPE, nodeName);
+        return new Authentication(user, authenticatedBy, null, Version.CURRENT, Authentication.AuthenticationType.TOKEN,
+            Map.of("_token_name", token.getTokenName()));
```

Comment:
Why the leading `_`?

Reply:
Not strictly necessary since we control service account entirely, but I think it is still better to use a reserved (with the leading `_`) name. It could be helpful on the client side for clarity.